### ISSUE TITLE
[CPU] MatMulNBits accuracy_level=4 (int8 activation quant) selects wrong argmax token on massive-activation LLMs (Qwen3-0.6B, Phi-3.5-mini)

### ISSUE THREAD

**Issue by justinchuby** [microsoft/onnxruntime#29849]
### Describe the issue

On the **CPU EP**, `com.microsoft::MatMulNBits` with **`accuracy_level=4`** (which int8-quantizes the **activation** `A` before an int8 matmul) can select a **different, incorrect argmax** for the output logits compared to `accuracy_level` 0/1/2/3, on transformer LLMs that have large-magnitude ("massive activation" / outlier) channels.

Because the ORT-GenAI model builder exports these models with `accuracy_level=4` **by default**, greedy decoding on CPU produces a **wrong token stream** relative to a higher-precision reference — silently. The first higher-precision level that agrees is `accuracy_level<=3`; only level 4 diverges.

We isolated this by building an fp32 oracle: take the exported `model.onnx`, rewrite **every** `MatMulNBits.accuracy_level` to `0`, and greedy-decode. Levels `0/1/2/3` all agree with each other; only the default `4` flips the token. The flip happens at **near-tie logits**, consistent with int8 activation quantization being too coarse for channels with large dynamic range.

### Two reproducible cases (fp32 oracle vs `accuracy_level=4`)

| Model (ORT-GenAI export) | Node width | Greedy token that flips | fp32 / acc 0-3 (correct) | acc=4 (default, wrong) |
|---|---|---|---|---|
| **Qwen3-0.6B** (`qwen3-0.6b-generic-cpu`) | has **8-bit** MatMulNBits | **decode index 0** (first token!) | `1479` | `3988` |
| **Phi-3.5-mini-instruct** (`generic-cpu`, int4 block-32) | 4-bit | decode index 65 | `263` | `6455` |

For Qwen3-0.6B the **very first generated token** is wrong under the default export. Both were reproduced independently by more than one person, and cross-checked: `accuracy_level` 1 (fp32), 2 (fp16), and 3 (bf16) all pick the correct token (`1479` / `263`); only `4` (int8 activation) picks the wrong one.

### Root-cause hypothesis

`accuracy_level=4` per-row/per-tensor int8-quantizes the activation vector `A` before the int8 GEMM. On these models a few channels carry very large magnitudes ("massive activations"), so an int8 activation scale loses the precision needed to resolve near-tie output logits, flipping the argmax. Keeping activations in fp32 (or int16) yields the correct token. Notably this bites the **8-bit-weight** nodes hardest (Qwen3-0.6B first token).

### Reproduction sketch

```python
import onnx, numpy as np, onnxruntime_genai as og  # or drive the graph directly

# 1) fp32 oracle: rewrite all MatMulNBits accuracy_level -> 0
m = onnx.load("model.onnx")   # keep external data
for node in m.graph.node:
    if node.op_type == "MatMulNBits":
        for attr in node.attribute:
            if attr.name == "accuracy_level":
                attr.i = 0
onnx.save(m, "model_acc0.onnx", save_as_external_data=True, location="model_acc0.onnx.data")

# 2) greedy-decode the SAME prompt with the original (accuracy_level=4) model and model_acc0.onnx
#    on the CPU EP, compare the generated token ids.
# Observed: they diverge — original picks the wrong token, acc0 matches the fp32 reference.
# (Repeat rewriting accuracy_level to 1/2/3 -> all agree with 0.)
```

Using ORT-GenAI or a raw greedy loop over the decoder graph both reproduce it. Prompt used for Phi-3.5 was a short "Hello"-style prompt; the divergence is deterministic (greedy).

### Urgency

Medium/High for correctness-sensitive CPU inference: default-exported quantized LLMs decode **incorrect greedy tokens** on CPU, differing from the fp32/fp16/bf16 reference, with no warning.

### Suggested direction

- Consider a higher-precision **int16 activation** path for `accuracy_level=4` (preserves the correct token in both cases above while staying faster than fp32), and/or
- per-channel activation scales for large-dynamic-range channels, and/or
- document that `accuracy_level=4` can change argmax on massive-activation models and pick a safer default for greedy/correctness-sensitive use.

### Platform / version

- onnxruntime **1.27.0**, CPU Execution Provider, Linux x86-64
- Models: ORT-GenAI model-builder exports of Qwen3-0.6B and Phi-3.5-mini-instruct (generic-cpu)

**Comment by blazingphoenix7** [comment]
I looked at this at the MLAS level on x64 (AVX2 plus AVX-VNNI, no AVX-512), built against `main` at d183c42. Short version: the int8 kernel is doing exactly what it is specified to do, but on an 8 bit node `accuracy_level=4` throws away more precision than the 8 bit weights buy, which is why those are the nodes that flip.

**One correction on the root cause note.** MLAS does not quantize `A` per row or per tensor. `QuantizeARow_CompInt8` in `sqnbitgemm_kernel_avx2.cpp` walks `A` in blocks of `BlkLen` along K and takes `scale = maxabs(block) / 127` per block, so activation scale granularity already matches the weight block granularity and is finer than per channel. A large channel only damages the one block of `BlkLen` values it sits in. Per channel activation scales would not buy anything on top of that.

**The kernel is not the problem.** I ran the real `MlasQNBitGemmBatch` at `SQNBIT_CompInt8` alongside an independent scalar reimplementation of per block int8 activation quantization accumulated in double, over 30 configurations (4 and 8 bit weights, `BlkLen` 16/32/64/128, outlier magnitude 1 to 1000). The kernel and the reference agree to every digit I printed in all 30, so this is the algorithm and not the implementation. `accuracy_level` 0 to 3 land where you would expect: `SQNBIT_CompFp32` sits between 5.3e-07 and 1.6e-05 rms against a double precision oracle over the same dequantized weights, with zero argmax flips in all 15 configurations where that kernel exists.

**Why the 8 bit nodes are the ones that break.** M=512, N=2048, K=1024, `BlkLen` 32, no outliers, rms logit error against a double precision GEMM on the original unquantized weights:

| weights | weight quant alone | weight quant plus int8 activations | int8 activation term on its own |
| --- | --- | --- | --- |
| 4 bit | 8.07e-02 | 8.09e-02 | 5.36e-03 |
| 8 bit | 4.75e-03 | 7.15e-03 | 5.35e-03 |

The int8 activation error is the same either way, because it does not depend on the weight bit width at all. On a 4 bit node it is 15x smaller than the weight error and is invisible. On an 8 bit node it is slightly larger than the weight error, so it becomes the single biggest error term in the matmul, bigger than the quantization the node was actually built for. That is the whole of the 8 bit observation. Worth noting that there is no `SQ8BitGemmVariant_CompFp32` in `qnbitgemm.cpp` at all, so on x64 an 8 bit node at level 0 to 3 does not go through QNBitGemm, it falls back to the operator dequantize path, which is why the jump at level 4 is so visible there.

**Massive activations scale that term, they do not create it.** With zero outliers the int8 path at `BlkLen` 32 still flipped 1.2 to 1.4 percent of argmaxes against the same dequantized weight oracle, so near tie logits are already exposed at baseline. Two outlier channels take the int8 activation rms from 5.4e-03 at ratio 1 to 2.2e-01 at ratio 1000. Block size is the multiplier: at ratio 100 the 4 bit error goes 4.1e-02 at `BlkLen` 16 to 1.2e-01 at `BlkLen` 128, and its argmax flip rate goes 2.0 to 4.7 percent, since one outlier crushes everything else sharing its block. Without outliers the same block size sweep only moves 4.7e-03 to 6.5e-03, so block size mostly matters once outliers are in play.

**On the int16 suggestion.** Simulating int16 activations at the same per block granularity drops the error about 256x (5.4e-03 to 2.1e-05 at `BlkLen` 32) and removed essentially every argmax flip. So the direction is right. The kernel cost is the catch: `VPDPBUSD` does 4 MACs per dword and `VPDPWSSD` does 2, so an int16 activation path is roughly half the int8 throughput before you count widening B.

Given the table, the cheapest fix is probably at export time rather than in MLAS. An 8 bit `MatMulNBits` node at `accuracy_level=4` spends more accuracy on the activations than it saved on the weights, so defaulting those to 4 in the model builder is close to self defeating, while 4 bit nodes are fine to leave there.

Caveat on the numbers: this is synthetic, gaussian activations with injected outlier channels, not your two models, so treat the flip percentages as illustrative and the error magnitudes as the actual result. I can put the harness up if that is useful, or prototype the int16 activation path in MLAS if you want a real measurement of what it costs.

**Comment by Nickgonzales76017** [comment]
I can contribute a narrow draft PR for the **model-level reproducibility/acceptance harness** around this defect, complementary to the synthetic MLAS harness already offered above. I maintain an exact-token quantization evaluator that records input identity, scored coverage, environment, checksums, and rejected integrity failures: https://github.com/Nickgonzales76017/bonfyre-fpq-benchmark

Proposed scope:

- add a standalone repro command that sweeps `accuracy_level=0..4` over one pinned Qwen3-0.6B export and one fixed token-ID prompt;
- record model/external-data hashes, rewritten node counts, provider/build identity, prompt IDs, generated token IDs, first-divergence index, reference/candidate logits, and argmax margin in canonical JSON;
- assert exact agreement for levels 0–3 and expose the level-4 divergence as an explicit expected-failure/acceptance condition until the default or kernel policy changes;
- fail closed on mixed model artifacts, incomplete external data, provider fallback, non-finite logits, unequal decode lengths, or missing rewrite coverage;
- add a tiny synthetic unit fixture for receipt determinism and validation while keeping the real model run scheduled/manual, so CI does not download or redistribute model weights.

I would keep the draft out of MLAS kernel design, exporter default selection, performance claims, model redistribution, and the Phi-3.5 case. It would establish one immutable Qwen3-0.6B correctness receipt that a later fix must turn green, without treating token agreement alone as a general quality claim.

If that test boundary is useful, I can inspect the current quantization-test/repro layout and open the draft PR.
